### PR TITLE
DAOS-11894 vos: Avoid segfault due to default evtree sort policy

### PR DESCRIPTION
**Pull request by jolivier23**
Modify default sort to use even split if the tree is too deep.  Otherwise, some I/O patterns can cause a segfault
due to the depth exceeding the EVT_TRACE_MAX

Rename ssof to soff for consistency

Required-githooks: true

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by NiuYawei — APPROVED** [review]

File: src/vos/evtree.c [hunk 1]
```diff
@@ -2131,6 +2131,21 @@ evt_large_hole_insert(daos_handle_t toh, const struct evt_entry_in *entry)
...
+{
+	/** EVT_FEAT_SORT_DIST can get imbalanced and, under certain conditions, can cause a
+	 *  segfault when the depth is larger than EVT_TRACE_MAX.  Rather than let it get too
+	 *  deep, let's detect cases where it starts getting deep and switch policies.
+	 */
+	if (unlikely(tcx->tc_depth > 5 &&
+		     (tcx->tc_feats & EVT_FEATS_SUPPORTED) == EVT_FEAT_SORT_DIST)) {
+		tcx->tc_feats ^= EVT_FEAT_SORT_DIST | EVT_FEAT_SORT_DIST_EVEN;
+		tcx->tc_ops = evt_policies[2];
+		D_DEBUG(DB_IO, "evtree depth is greater than 10, switching sort policy\n");
```
Comment (Nasf-Fan) [inline src/vos/evtree.c:2145]: 5 or 10? "greater than tcx->tc_depth" to avoid misguiding.

File: src/vos/evtree.c [hunk 2]
```diff
@@ -2131,6 +2131,21 @@ evt_large_hole_insert(daos_handle_t toh, const struct evt_entry_in *entry)
...
 
+static inline void
+evt_check_sort_policy(struct evt_context *tcx)
+{
+	/** EVT_FEAT_SORT_DIST can get imbalanced and, under certain conditions, can cause a
+	 *  segfault when the depth is larger than EVT_TRACE_MAX.  Rather than let it get too
+	 *  deep, let's detect cases where it starts getting deep and switch policies.
+	 */
+	if (unlikely(tcx->tc_depth > 5 &&
+		     (tcx->tc_feats & EVT_FEATS_SUPPORTED) == EVT_FEAT_SORT_DIST)) {
```
Comment (Nasf-Fan) [inline src/vos/evtree.c:2142]: I am afraid that only checking the depth may be not enough to say we are in very imbalance status, the total records count seems another importance factor (but seems we do not maintain that). If it is not imbalance issue caused depth, then what is the overhead for such switch?

File: src/vos/evtree.c (hunk 1 above)
Comment (jolivier23) [reply src/vos/evtree.c:2145]: Good point. I initially picked 10.

File: src/vos/evtree.c (hunk 2 above)
Comment (jolivier23) [reply src/vos/evtree.c:2142]: If the tree depth is 6 (23^6 / 2), we can accommodate ~70 million entries in a balanced tree.  I think it's fairly unlikely to hit this without significant imbalance under any normal use case.   For instance, in the array API, we typically only write one stripe to a given evtree.   The ticket is kind of a worst case scenario and would only result in about 1 million entries in the tree, assuming no aggregation to clean up old entries.

Comment (jolivier23) [reply src/vos/evtree.c:2143]: I changed it to 6.   In my testing, I don't see other patterns manifesting deep tree growth.  In other words, it's really just this type of I/O patterns that causes the default sort to misbehave.

**Review by NiuYawei — DISMISSED** [review]

**Review by Nasf-Fan — DISMISSED** [review]

**Review by NiuYawei — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

File: src/vos/evtree.c [hunk 3]
```diff
@@ -93,18 +93,18 @@ enum {
 	RT_OVERLAP_PARTIAL	= (1 << 6),
 };
 
-static struct evt_policy_ops evt_ssof_pol_ops;
+static struct evt_policy_ops  evt_soff_pol_ops;
```
Comment (ryon-jensen) [inline src/vos/evtree.c:96]: nit - extra space added

**Review by ryon-jensen — APPROVED** [review]

**Comment by jolivier23** [comment]
Failure due to DAOS-11944, DAOS-11951

File: src/vos/evtree.c (hunk 3 above)
Comment (jolivier23) [reply src/vos/evtree.c:96]: That might have been clang-format

**Pull request merged**